book/continuous_integration.md

```diff
@@ -0,0 +1,192 @@
+
+## Pre-Commit and git hooks
+
+An alternative to continuous integration, is using a Git hook.
+[Git hooks](https://git-scm.com/docs/githooks) are scripts that can be set to run locally at specific points in your Git workflow.
+For example, we might set up a `pre-commit` or `pre-push` hook that runs our tests before we make each commit or push to the remote repository.
+This might stop our commit/push if the tests fail, so that we don't push breaking changes to our remote repository.
+
+```{note}
+If your code is likely to be run on a range of software versions or operating systems, you might want to test on a variety of these. Tools exists to support local testing of combinations software versions and package dependency versions:
+* [tox](https://tox.readthedocs.io/en/latest/) or [nox](https://nox.thea.codes/en/stable/) for Python
+* [rhub](https://r-hub.github.io/rhub/) for R
+
+However, [continuous integration](continuous-integration) can be used to automate these tests on a broader range on parameters.
+```
+
+(linters-formatters)=
+## Linters and formatters
+
+As discussed in [](automate-style-checks), the process of checking and fixing code for style and formatting is tedious. Automation can speed up this work, either by providing suggestions as the code is written or by reformatting your code to comply with some style.
+
+Two main types of tool exist for these tasks:
+
+* Linters - these analyse your code to flag stylistic errors (and sometimes bugs or security issues too).
+* Formatters - these not only detect when you have diverged from a style, but will automatically correct the formatting of your code to conform to a particular style.
+
+```{list-table} Packages that can be used for linting or formatting in Python and R
+:header-rows: 1
+:name: linters
+
+* - Language
+  - Linters
+  - Formatters
+* - Python
+  - `flake8`, `pylint`, `Bandit`
+  - `Black`, `Isort`
+* - R
+  - `lintr`
+  - `formatR`, `styler`
+```
+
+Be sure to read the documentation for any of these tools, to understand what they are checking or changing in your code. Some can be configured to ignore or detect specific types of formatting error. You can run multiple of these, to catch a broader range of stylistic or programmatic errors.
+
+## Example use cases for GitHub Actions
+
+The following examples are presented to support understanding of the topic discussed above.
+
+### Configure GitHub actions to automate tests
+
+Below is an example configuration file, for use with GitHub actions. The `YAML` file format, used below, is common to a number of other CI tools.
+
+```yaml
+name: Test python package
+
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+
+jobs:
+  build:
+
+    runs-on: ubuntu-latest
```

Comment:
The two workflow examples you have show how to use a single python version, and how to run over multiple which is great for users to test over a range of python versions. Is it worth doing the same for the operating systems so users can see how to do both?

1st example:
multiple os's and python versions

2nd example
1 os and 1 python version.

Idea:
runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        os: [ ubuntu-latest, macos-latest, windows-latest ]
        python: [ 3.8, 3.9, '3.10', 3.11 ]

Reply:
yes, I think that's a great idea.

Reply:
@joshlynchONS will find some further examples and add links to these. We want to focus on guidance so will provide links rather than expand the examples. Although having the examples has proved incredibly helpful to Josh and me.